Makefile

```diff
@@ -143,59 +241,143 @@ deploy: manifests kustomize prepare-local-clusterctl docker-push-kind ## Deploy
 
 .PHONY: prepare-local-clusterctl
 prepare-local-clusterctl: manifests kustomize  ## Prepare overide file for local clusterctl.
 	mkdir -p ~/.cluster-api/overrides/infrastructure-nutanix/${LOCAL_PROVIDER_VERSION}
 	cd config/manager && $(KUSTOMIZE) edit set image controller=${IMG}
 	$(KUSTOMIZE) build config/default > ~/.cluster-api/overrides/infrastructure-nutanix/${LOCAL_PROVIDER_VERSION}/infrastructure-components.yaml
 	cp ./metadata.yaml ~/.cluster-api/overrides/infrastructure-nutanix/${LOCAL_PROVIDER_VERSION}
-	cp ./cluster-template.yaml ~/.cluster-api/overrides/infrastructure-nutanix/${LOCAL_PROVIDER_VERSION}
+	cp ./templates/cluster-template.yaml ~/.cluster-api/overrides/infrastructure-nutanix/${LOCAL_PROVIDER_VERSION}
 	cp ./clusterctl.yaml ~/.cluster-api/clusterctl.yaml
 
-CONTROLLER_GEN = $(shell pwd)/bin/controller-gen
-.PHONY: controller-gen
-controller-gen: ## Download controller-gen locally if necessary.
-	$(call go-get-tool,$(CONTROLLER_GEN),sigs.k8s.io/controller-tools/cmd/controller-gen@v0.8.0)
-
-CONVERSION_GEN = $(shell pwd)/bin/conversion-gen
-.PHONY: conversion-gen
-conversion-gen: ## Download conversion-gen locally if necessary.
-	rm -f $(CONVERSION_GEN)
-	$(call go-get-tool,$(CONVERSION_GEN),k8s.io/code-generator/cmd/conversion-gen@v0.23.6)
-
-KUSTOMIZE = $(shell pwd)/bin/kustomize
-.PHONY: kustomize
-kustomize: ## Download kustomize locally if necessary.
-	$(call go-get-tool,$(KUSTOMIZE),sigs.k8s.io/kustomize/kustomize/v4@v4.5.4)
-
-ENVTEST = $(shell pwd)/bin/setup-envtest
-.PHONY: envtest
-envtest: ## Download envtest-setup locally if necessary.
-	$(call go-get-tool,$(ENVTEST),sigs.k8s.io/controller-runtime/tools/setup-envtest@latest)
-
-KO = $(shell pwd)/bin/ko
-.PHONY: ko
-ko: ## Download ko locally if necessary.
-	$(call go-get-tool,$(KO),github.com/google/ko@latest)
-
-# go-get-tool will 'go get' any package $2 and install it to $1.
-PROJECT_DIR := $(shell dirname $(abspath $(lastword $(MAKEFILE_LIST))))
-define go-get-tool
-@[ -f $(1) ] || { \
-set -e ;\
-echo "Downloading $(2)" ;\
-GOBIN=$(PROJECT_DIR)/bin go install $(2) ;\
-}
-endef
+.PHONY: test-unittest
+test-unittest: manifests generate fmt vet envtest ## Run unit tests.
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION)  --arch=amd64 -p path)" go test ./... -coverprofile cover.out
+
+.PHONY: test-clusterctl
+test-clusterctl: prepare-local-clusterctl ## Run the tests using clusterctl
+	which clusterctl
+	clusterctl version
+	clusterctl config repositories | grep nutanix
+	clusterctl generate cluster ${TEST_CLUSTER_NAME} -i nutanix:${LOCAL_PROVIDER_VERSION} --list-variables -v 10
+	clusterctl generate cluster ${TEST_CLUSTER_NAME} -i nutanix:${LOCAL_PROVIDER_VERSION} --target-namespace ${TEST_NAMESPACE}  -v 10 > ./cluster.yaml
+	kubectl create ns $(TEST_NAMESPACE) || true
+	kubectl apply -f ./cluster.yaml -n $(TEST_NAMESPACE)
+
+.PHONY: test-e2e
+test-e2e: docker-build-e2e $(GINKGO) cluster-templates ## Run the end-to-end tests
```

Comment:
$(KO) is missing from test-e2e

Reply:
one can do <pre>make ko</pre>
did not add here to avoid repeated checks.

Reply:
will add it to the docker-build-e2e section